File: backend/services/vision/deterministic.py
"""Model-free image measurements.

Adapted from macchant/imex-next (MIT) — `pipeline/color.ts` and
`pipeline/analyze.ts`: CIELAB k-means palette, border-based background
isolation, vector-likeness, Sobel edge density, aspect snapping and EXIF.
Ported to numpy/PIL; the TypeScript twin in `frontend/lib/shotspec/
deterministic.ts` gives the renderer an instant preview of the same numbers.
Server values are the ones stored in the ShotSpec.
"""

from __future__ import annotations

import math
from pathlib import Path

import numpy as np
from numpy.typing import NDArray
from PIL import Image, ImageOps

from services.vision.protocol import MeasuredStats, PaletteEntry

_ANALYSIS_EDGE = 256
_PALETTE_K = 6
_ASPECTS: dict[str, float] = {
    "1:1": 1.0,
    "4:3": 4 / 3,
    "3:2": 3 / 2,
    "16:9": 16 / 9,
    "21:9": 21 / 9,
    "9:16": 9 / 16,
    "2:3": 2 / 3,
    "3:4": 3 / 4,
    "4:5": 4 / 5,
    "5:4": 5 / 4,
}
_EXIF_TAGS: dict[int, str] = {
    271: "make",
    272: "model",
    306: "datetime",
    33434: "exposure_time",
    33437: "f_number",
    34855: "iso",
    37386: "focal_length",
    41989: "focal_length_35mm",
}

FloatArray = NDArray[np.float64]


def snap_aspect(width: int, height: int) -> str:
    ratio = width / height if height else 1.0
    best = min(_ASPECTS.items(), key=lambda item: abs(math.log(ratio / item[1])))
    return best[0] if abs(math.log(ratio / best[1])) < 0.08 else f"{width}:{height}"


def _srgb_to_lab(rgb: FloatArray) -> FloatArray:
    """rgb in 0..1, shape (n, 3) → CIELAB (D65)."""
    linear = np.where(rgb <= 0.04045, rgb / 12.92, ((rgb + 0.055) / 1.055) ** 2.4)
    matrix = np.array([[0.4124564, 0.3575761, 0.1804375], [0.2126729, 0.7151522, 0.0721750], [0.0193339, 0.1191920, 0.9503041]])
    xyz = linear @ matrix.T
    xyz /= np.array([0.95047, 1.0, 1.08883])
    f = np.where(xyz > 0.008856, np.cbrt(xyz), 7.787 * xyz + 16 / 116)
    lab = np.stack([116 * f[:, 1] - 16, 500 * (f[:, 0] - f[:, 1]), 200 * (f[:, 1] - f[:, 2])], axis=1)
    return lab


def _lab_to_hex(lab: FloatArray) -> str:
    l, a, b = float(lab[0]), float(lab[1]), float(lab[2])
    fy = (l + 16) / 116
    fx = fy + a / 500
    fz = fy - b / 200
    def inv(t: float) -> float:
        return t**3 if t**3 > 0.008856 else (t - 16 / 116) / 7.787
    x, y, z = inv(fx) * 0.95047, inv(fy), inv(fz) * 1.08883
    matrix = np.array([[3.2404542, -1.5371385, -0.4985314], [-0.9692660, 1.8760108, 0.0415560], [0.0556434, -0.2040259, 1.0572252]])
    linear = matrix @ np.array([x, y, z])
    srgb = np.where(linear <= 0.0031308, 12.92 * linear, 1.055 * np.power(np.clip(linear, 0, None), 1 / 2.4) - 0.055)
    r, g, bb = (int(round(float(v) * 255)) for v in np.clip(srgb, 0, 1))
    return f"#{r:02x}{g:02x}{bb:02x}"


def kmeans_palette(rgb: FloatArray, k: int = _PALETTE_K, iterations: int = 12, seed: int = 7) -> list[PaletteEntry]:
    """k-means in CIELAB so cluster distances match perceived colour distance.
    Deterministic: fixed seed, fixed iteration count, largest cluster first."""
    if rgb.shape[0] == 0:
        return []
    lab = _srgb_to_lab(rgb)
    rng = np.random.default_rng(seed)
    k = min(k, lab.shape[0])
    centers = lab[rng.choice(lab.shape[0], size=k, replace=False)]
    labels = np.zeros(lab.shape[0], dtype=np.int64)
    for _ in range(iterations):
        distances = ((lab[:, None, :] - centers[None, :, :]) ** 2).sum(axis=2)
        labels = distances.argmin(axis=1)
        for index in range(k):
            members = lab[labels == index]
            if members.shape[0]:
                centers[index] = members.mean(axis=0)
    counts = np.bincount(labels, minlength=k).astype(np.float64)
    order = np.argsort(-counts)
    total = float(counts.sum()) or 1.0
    entries: list[PaletteEntry] = []
    for index in order:
        if counts[index] == 0:
            continue
        entries.append(PaletteEntry(hex=_lab_to_hex(centers[index]), share=round(float(counts[index] / total), 4)))
    return entries


def _sobel_edges(gray: FloatArray) -> FloatArray:
    kx = np.array([[-1, 0, 1], [-2, 0, 2], [-1, 0, 1]], dtype=np.float64)
    ky = kx.T
    padded = np.pad(gray, 1, mode="edge")
    gx = np.zeros_like(gray)
    gy = np.zeros_like(gray)
    for dy in range(3):
        for dx in range(3):
            window = padded[dy : dy + gray.shape[0], dx : dx + gray.shape[1]]
            gx += kx[dy, dx] * window
            gy += ky[dy, dx] * window
    return np.hypot(gx, gy)


def _laplacian_variance(gray: FloatArray) -> float:
    padded = np.pad(gray, 1, mode="edge")
    center = padded[1:-1, 1:-1]
    lap = padded[:-2, 1:-1] + padded[2:, 1:-1] + padded[1:-1, :-2] + padded[1:-1, 2:] - 4 * center
    return float(lap.var())


def _exif(image: Image.Image) -> dict[str, str]:
    result: dict[str, str] = {}
    try:
        exif = image.getexif()
    except Exception:  # noqa: BLE001 - EXIF is optional
        return result
    for tag, name in _EXIF_TAGS.items():
        value = exif.get(tag)
        if value is None:
            continue
        result[name] = str(value)
    try:
        detail = exif.get_ifd(0x8769)
        for tag, name in _EXIF_TAGS.items():
            value = detail.get(tag)
            if value is not None and name not in result:
                result[name] = str(value)
    except Exception:  # noqa: BLE001
        pass
    return result


def measure_image(image: Image.Image) -> MeasuredStats:
    width, height = image.size
    exif = _exif(image)
    oriented = ImageOps.exif_transpose(image) or image
    rgb_image = oriented.convert("RGB")
    rgb_image.thumbnail((_ANALYSIS_EDGE, _ANALYSIS_EDGE))
    rgb = np.asarray(rgb_image, dtype=np.float64) / 255.0
    flat = rgb.reshape(-1, 3)
    gray = 0.2126 * rgb[:, :, 0] + 0.7152 * rgb[:, :, 1] + 0.0722 * rgb[:, :, 2]
    hsv = np.asarray(rgb_image.convert("HSV"), dtype=np.float64) / 255.0
    saturation = float(hsv[:, :, 1].mean())
    luminance = float(gray.mean())
    contrast = float(gray.std())
    edges = _sobel_edges(gray)
    edge_density = float((edges > 0.25).mean())
    sharpness = _laplacian_variance(gray)

    # Background: the border ring; a "clear" background means the ring is one colour.
    ring = np.concatenate([rgb[0, :, :], rgb[-1, :, :], rgb[:, 0, :], rgb[:, -1, :]], axis=0)
    ring_lab = _srgb_to_lab(ring)
    ring_spread = float(ring_lab.std(axis=0).mean())
    background_hex = _lab_to_hex(ring_lab.mean(axis=0)) if ring_spread < 6.0 else ""

    # Vector-likeness: few distinct colours + hard edges.
    quantised = (flat * 15).round().astype(np.int64)
    unique_colours = np.unique(quantised, axis=0).shape[0]
    colour_term = max(0.0, 1.0 - unique_colours / 400.0)
    hard_edge_term = float((edges > 1.0).sum() / max(1.0, (edges > 0.25).sum()))
    vector_likeness = round(min(1.0, 0.6 * colour_term + 0.4 * hard_edge_term), 3)

    palette = kmeans_palette(flat)
    return MeasuredStats(
        width=width,
        height=height,
        aspect=snap_aspect(width, height),
        palette=palette,
        luminance=round(luminance, 4),
        contrast=round(contrast, 4),
        saturation=round(saturation, 4),
        edge_density=round(edge_density, 4),
        sharpness=round(sharpness, 6),
        background_hex=background_hex,
        vector_likeness=vector_likeness,
        exif=exif,
    )


def measure_path(path: str | Path) -> MeasuredStats:
    with Image.open(path) as image:
        return measure_image(image)

File: frontend/lib/shotspec/deterministic.ts
/**
 * Model-free image measurements — the TypeScript twin of
 * `backend/services/vision/deterministic.py` (both adapted from
 * macchant/imex-next, MIT: `pipeline/color.ts`, `pipeline/analyze.ts`).
 *
 * The renderer uses it for an instant preview while the server computes the
 * authoritative numbers that go into the ShotSpec. Same algorithm, same
 * constants, same seed, so the preview and the stored values agree.
 */

export interface PaletteEntry {
  hex: string
  share: number
}

export interface MeasuredStats {
  width: number
  height: number
  aspect: string
  palette: PaletteEntry[]
  luminance: number
  contrast: number
  saturation: number
  edge_density: number
  sharpness: number
  background_hex: string
  vector_likeness: number
}

const ANALYSIS_EDGE = 256
const PALETTE_K = 6
const ASPECTS: [string, number][] = [
  ['1:1', 1],
  ['4:3', 4 / 3],
  ['3:2', 3 / 2],
  ['16:9', 16 / 9],
  ['21:9', 21 / 9],
  ['9:16', 9 / 16],
  ['2:3', 2 / 3],
  ['3:4', 3 / 4],
  ['4:5', 4 / 5],
  ['5:4', 5 / 4],
]

export function snapAspect(width: number, height: number): string {
  const ratio = height ? width / height : 1
  let best: [string, number] = ASPECTS[0]
  let bestDistance = Infinity
  for (const entry of ASPECTS) {
    const distance = Math.abs(Math.log(ratio / entry[1]))
    if (distance < bestDistance) {
      bestDistance = distance
      best = entry
    }
  }
  return bestDistance < 0.08 ? best[0] : `${width}:${height}`
}

function srgbToLab(r: number, g: number, b: number): [number, number, number] {
  const lin = (c: number) => (c <= 0.04045 ? c / 12.92 : ((c + 0.055) / 1.055) ** 2.4)
  const rl = lin(r), gl = lin(g), bl = lin(b)
  let x = (0.4124564 * rl + 0.3575761 * gl + 0.1804375 * bl) / 0.95047
  let y = 0.2126729 * rl + 0.7151522 * gl + 0.072175 * bl
  let z = (0.0193339 * rl + 0.119192 * gl + 0.9503041 * bl) / 1.08883
  const f = (t: number) => (t > 0.008856 ? Math.cbrt(t) : 7.787 * t + 16 / 116)
  x = f(x); y = f(y); z = f(z)
  return [116 * y - 16, 500 * (x - y), 200 * (y - z)]
}

function labToHex(l: number, a: number, b: number): string {
  const fy = (l + 16) / 116
  const fx = fy + a / 500
  const fz = fy - b / 200
  const inv = (t: number) => (t ** 3 > 0.008856 ? t ** 3 : (t - 16 / 116) / 7.787)
  const x = inv(fx) * 0.95047, y = inv(fy), z = inv(fz) * 1.08883
  const lr = 3.2404542 * x - 1.5371385 * y - 0.4985314 * z
  const lg = -0.969266 * x + 1.8760108 * y + 0.041556 * z
  const lb = 0.0556434 * x - 0.2040259 * y + 1.0572252 * z
  const gamma = (c: number) => {
    const v = c <= 0.0031308 ? 12.92 * c : 1.055 * Math.pow(Math.max(0, c), 1 / 2.4) - 0.055
    return Math.round(Math.min(1, Math.max(0, v)) * 255)
  }
  const toHex = (v: number) => v.toString(16).padStart(2, '0')
  return `#${toHex(gamma(lr))}${toHex(gamma(lg))}${toHex(gamma(lb))}`
}

/** numpy's PCG64 is not reproducible here; a small deterministic LCG picks the seeds instead. */
function makeRng(seed: number): () => number {
  let state = seed >>> 0
  return () => {
    state = (Math.imul(state, 1664525) + 1013904223) >>> 0
    return state / 0x100000000
  }
}

/** k-means in CIELAB; deterministic (fixed seed, fixed iterations), largest cluster first. */
export function kmeansPalette(rgb: Float64Array, k = PALETTE_K, iterations = 12, seed = 7): PaletteEntry[] {
  const n = rgb.length / 3
  if (n === 0) return []
  const lab = new Float64Array(n * 3)
  for (let i = 0; i < n; i++) {
    const [l, a, b] = srgbToLab(rgb[i * 3], rgb[i * 3 + 1], rgb[i * 3 + 2])
    lab[i * 3] = l; lab[i * 3 + 1] = a; lab[i * 3 + 2] = b
  }
  k = Math.min(k, n)
  const rng = makeRng(seed)
  const chosen = new Set<number>()
  while (chosen.size < k) chosen.add(Math.floor(rng() * n))
  const centers = Array.from(chosen, i => [lab[i * 3], lab[i * 3 + 1], lab[i * 3 + 2]])
  const labels = new Int32Array(n)
  for (let iter = 0; iter < iterations; iter++) {
    for (let i = 0; i < n; i++) {
      let best = 0, bestD = Infinity
      for (let c = 0; c < k; c++) {
        const d = (lab[i * 3] - centers[c][0]) ** 2 + (lab[i * 3 + 1] - centers[c][1]) ** 2 + (lab[i * 3 + 2] - centers[c][2]) ** 2
        if (d < bestD) { bestD = d; best = c }
      }
      labels[i] = best
    }
    const sums = centers.map(() => [0, 0, 0, 0])
    for (let i = 0; i < n; i++) {
      const s = sums[labels[i]]
      s[0] += lab[i * 3]; s[1] += lab[i * 3 + 1]; s[2] += lab[i * 3 + 2]; s[3] += 1
    }
    for (let c = 0; c < k; c++) if (sums[c][3] > 0) centers[c] = [sums[c][0] / sums[c][3], sums[c][1] / sums[c][3], sums[c][2] / sums[c][3]]
  }
  const counts = new Array<number>(k).fill(0)
  for (let i = 0; i < n; i++) counts[labels[i]]++
  const order = counts.map((count, index) => [count, index]).sort((a, b) => b[0] - a[0])
  const total = n || 1
  return order.filter(([count]) => count > 0).map(([count, index]) => ({ hex: labToHex(centers[index][0], centers[index][1], centers[index][2]), share: Math.round((count / total) * 10000) / 10000 }))
}

function sobel(gray: Float64Array, width: number, height: number): Float64Array {
  const out = new Float64Array(width * height)
  const at = (x: number, y: number) => gray[Math.min(height - 1, Math.max(0, y)) * width + Math.min(width - 1, Math.max(0, x))]
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const gx = -at(x - 1, y - 1) + at(x + 1, y - 1) - 2 * at(x - 1, y) + 2 * at(x + 1, y) - at(x - 1, y + 1) + at(x + 1, y + 1)
      const gy = -at(x - 1, y - 1) - 2 * at(x, y - 1) - at(x + 1, y - 1) + at(x - 1, y + 1) + 2 * at(x, y + 1) + at(x + 1, y + 1)
      out[y * width + x] = Math.hypot(gx, gy)
    }
  }
  return out
}

function laplacianVariance(gray: Float64Array, width: number, height: number): number {
  const at = (x: number, y: number) => gray[Math.min(height - 1, Math.max(0, y)) * width + Math.min(width - 1, Math.max(0, x))]
  const values = new Float64Array(width * height)
  let mean = 0
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const v = at(x, y - 1) + at(x, y + 1) + at(x - 1, y) + at(x + 1, y) - 4 * at(x, y)
      values[y * width + x] = v
      mean += v
    }
  }
  mean /= values.length || 1
  let variance = 0
  for (const v of values) variance += (v - mean) ** 2
  return variance / (values.length || 1)
}

/** Nearest-neighbour downscale of RGBA pixels so the analysis matches the server's 256 px thumbnail. */
function downscale(rgba: Uint8ClampedArray, width: number, height: number): { rgb: Float64Array; w: number; h: number } {
  const scale = Math.min(1, ANALYSIS_EDGE / Math.max(width, height))
  const w = Math.max(1, Math.round(width * scale))
  const h = Math.max(1, Math.round(height * scale))
  const rgb = new Float64Array(w * h * 3)
  for (let y = 0; y < h; y++) {
    const sy = Math.min(height - 1, Math.floor((y + 0.5) / scale))
    for (let x = 0; x < w; x++) {
      const sx = Math.min(width - 1, Math.floor((x + 0.5) / scale))
      const i = (sy * width + sx) * 4
      const o = (y * w + x) * 3
      rgb[o] = rgba[i] / 255; rgb[o + 1] = rgba[i + 1] / 255; rgb[o + 2] = rgba[i + 2] / 255
    }
  }
  return { rgb, w, h }
}

/** Measure raw RGBA pixels (what `canvas.getImageData` returns). Pure; usable in tests without a DOM. */
export function measurePixels(rgba: Uint8ClampedArray, width: number, height: number): MeasuredStats {
  const { rgb, w, h } = downscale(rgba, width, height)
  const n = w * h
  const gray = new Float64Array(n)
  let lumSum = 0, satSum = 0
  for (let i = 0; i < n; i++) {
    const r = rgb[i * 3], g = rgb[i * 3 + 1], b = rgb[i * 3 + 2]
    gray[i] = 0.2126 * r + 0.7152 * g + 0.0722 * b
    lumSum += gray[i]
    const max = Math.max(r, g, b), min = Math.min(r, g, b)
    satSum += max === 0 ? 0 : (max - min) / max
  }
  const luminance = lumSum / n
  let varSum = 0
  for (let i = 0; i < n; i++) varSum += (gray[i] - luminance) ** 2
  const contrast = Math.sqrt(varSum / n)
  const edges = sobel(gray, w, h)
  let edgeCount = 0, hardCount = 0
  for (const e of edges) {
    if (e > 0.25) edgeCount++
    if (e > 1.0) hardCount++
  }
  const edgeDensity = edgeCount / n
  const sharpness = laplacianVariance(gray, w, h)

  // Background from the border ring.
  const ringLab: [number, number, number][] = []
  const pushLab = (x: number, y: number) => {
    const i = (y * w + x) * 3
    ringLab.push(srgbToLab(rgb[i], rgb[i + 1], rgb[i + 2]))
  }
  for (let x = 0; x < w; x++) { pushLab(x, 0); pushLab(x, h - 1) }
  for (let y = 0; y < h; y++) { pushLab(0, y); pushLab(w - 1, y) }
  const mean = [0, 0, 0]
  for (const l of ringLab) { mean[0] += l[0]; mean[1] += l[1]; mean[2] += l[2] }
  mean[0] /= ringLab.length; mean[1] /= ringLab.length; mean[2] /= ringLab.length
  const spread = [0, 0, 0]
  for (const l of ringLab) { spread[0] += (l[0] - mean[0]) ** 2; spread[1] += (l[1] - mean[1]) ** 2; spread[2] += (l[2] - mean[2]) ** 2 }
  const ringSpread = (Math.sqrt(spread[0] / ringLab.length) + Math.sqrt(spread[1] / ringLab.length) + Math.sqrt(spread[2] / ringLab.length)) / 3
  const backgroundHex = ringSpread < 6 ? labToHex(mean[0], mean[1], mean[2]) : ''

  const unique = new Set<number>()
  for (let i = 0; i < n; i++) unique.add((Math.round(rgb[i * 3] * 15) << 8) | (Math.round(rgb[i * 3 + 1] * 15) << 4) | Math.round(rgb[i * 3 + 2] * 15))
  const colourTerm = Math.max(0, 1 - unique.size / 400)
  const hardEdgeTerm = hardCount / Math.max(1, edgeCount)
  const vectorLikeness = Math.round(Math.min(1, 0.6 * colourTerm + 0.4 * hardEdgeTerm) * 1000) / 1000

  const round4 = (v: number) => Math.round(v * 10000) / 10000
  return {
    width,
    height,
    aspect: snapAspect(width, height),
    palette: kmeansPalette(rgb),
    luminance: round4(luminance),
    contrast: round4(contrast),
    saturation: round4(satSum / n),
    edge_density: round4(edgeDensity),
    sharpness: Math.round(sharpness * 1e6) / 1e6,
    background_hex: backgroundHex,
    vector_likeness: vectorLikeness,
  }
}

/** Measure a loaded image element via an offscreen canvas. */
export function measureImageElement(image: HTMLImageElement): MeasuredStats {
  const canvas = document.createElement('canvas')
  const scale = Math.min(1, ANALYSIS_EDGE / Math.max(image.naturalWidth, image.naturalHeight))
  canvas.width = Math.max(1, Math.round(image.naturalWidth * scale))
  canvas.height = Math.max(1, Math.round(image.naturalHeight * scale))
  const context = canvas.getContext('2d')
  if (!context) throw new Error('2D canvas unavailable')
  context.drawImage(image, 0, 0, canvas.width, canvas.height)
  const data = context.getImageData(0, 0, canvas.width, canvas.height).data
  const stats = measurePixels(data, canvas.width, canvas.height)
  return { ...stats, width: image.naturalWidth, height: image.naturalHeight, aspect: snapAspect(image.naturalWidth, image.naturalHeight) }
}
